crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,248 @@
+use crate::price_estimation::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::Result;
+use futures::future::Shared;
+use futures::FutureExt;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+type SharedEstimationRequest = Shared<Pin<Box<dyn Future<Output = EstimationResult> + Send>>>;
+
+struct Inner {
+    estimator: Box<dyn PriceEstimating>,
+    in_flight_requests: Mutex<HashMap<Query, SharedEstimationRequest>>,
+}
+
+impl Inner {
+    fn collect_garbage(&self) {
+        // TODO: Refactor this to use `HashMap::drain_filter` when it's stable:
+        // https://github.com/rust-lang/rust/issues/59618
+        let mut active_requests = self.in_flight_requests.lock().unwrap();
+        let completed_and_ignored_requests: Vec<_> = active_requests
+            .iter()
+            .filter_map(|(query, handle)| {
+                if matches!(handle.strong_count(), Some(1)) {
+                    // Only `Inner::active_requests` is still holding on to it.
+                    Some(*query)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        for query in &completed_and_ignored_requests {
+            active_requests.remove(query);
+        }
+    }
+}
+
+/// A price estimator which doesn't issue another estimation request while an identical one is
+/// already in-flight.
+pub struct BufferingPriceEstimator {
+    inner: Arc<Inner>,
+}
+
+impl BufferingPriceEstimator {
+    pub fn new(estimator: Box<dyn PriceEstimating>) -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                estimator,
+                in_flight_requests: Mutex::new(Default::default()),
+            }),
+        }
+    }
+
+    async fn estimate_buffered(&self, queries: &[Query]) -> Vec<EstimationResult> {
+        // For each `Query` either get an in-flight request or keep the `Query` to forward it to the
+        // inner price estimator.
+        let (active_requests, remaining_queries): (Vec<_>, Vec<_>) = {
+            let requests = self.inner.in_flight_requests.lock().unwrap();
```

Comment:
Should this lock be held until line 111 where in_flight_requests has been extended? During this time we do not perform any async operations. Not holding the lock allows the following to happen:
```
thread 1 reaches this line, locks the mutex, determines it needs to create a new future, unlocks the mutex
thread 2 reaches this line, locks the mutex, determines it needs to create a new future, unlocks the mutex
thread 1 reaches line 102, locks the mutex, inserts its future as in flight, unlocks the mutex
thread 2 reaches line 102, locks the mutex, inserts its future as in flight, unlocks the mutex
```
In this situation we have duplicate in flight requests which could have been prevented.

Reply:
I initially held the lock while creating the futures but preferred to have a shorted critical section with the downside of possibly allowing duplicated requests.
I would be fine with either version.

Reply:
To me it makes sense because we're just creating a couple of vectors in between and this completely guarantees that there won't be duplicate requests in flight at the same time.
It might even be faster overall because there is only one lock instead of two.

Reply:
Alright. Then I will extend the lock accordingly. 👍

Reply:
This also makes sense to me.